### ISSUE TITLE
traffic from wlan want through ss-redir but not properly NATed 

### ISSUE THREAD

**Issue by blurrcat** [shadowsocks/shadowsocks-libev#1844]
I know this is probably not a problem with shadowsocks-libev itself. But I couldn't find answers to it after LOTS of searches, mainly due to my lack of knowledge to iptables.. I figure this can be helpful to others if solved since it's a common use case.

Any help is appreciated, thanks in advance!

### What version of shadowsocks-libev are you using?

Manually built, commit 3ed6019

### What operating system are you using?

Fedora-Minimal-27-1.6.aarch64 on Raspberry Pi 3

### What did you do?

Since the official fedora repo doesn't contain a prebuilt for arm, I followed the instructions "Building from source with centos". The build was successful and I'm able to run ss-redir and ss-nat.

The setup: My Pi is connected to the home router on eth0(192.168.1.11) and I used NetworkManager to create a wifi hotspot(wlan0, 10.42.1.0/24), so other device can connect to pi via wifi and share the internet. Sth like the following:

    | home router | ---- | pi:eth0(192.168.1.11) - pi:wlan0(10.42.1.1) | ---- | android phone(10.42.1.13) |

The following rule is added by NetworkManager for traffic from wlan0:

    -A POSTROUTING -s 10.42.0.0/24 ! -d 10.42.0.0/24 -j MASQUERADE

I have ss-redir running and used ss-nat to setup iptable rules. I can verify this is working on Pi(`curl https://goo.gl` ok).

The commands used are listed below:

    # create hotspot
    nmcli dev wifi hotspot ifname wlan0 con-name wlan0 ...
    # run ss-redir
    ss-redir -c config.json --fast-open
    # update iptables
    ss-nat -s xx.xx.xx.xx -l 1080 -o -i chinaroutes.txt -I eth0  # also tried -I wlan0, not working
    
### What did you expect to see?

I wish traffic from other devices connected through wifi can be routed through ss-redir. In the example above, the android device should be able to access to https://goo.gl.

### What did you see instead?

Traffic destined to china routes work fine. Traffic that should go through ss-dir seems to be dropped.
I'm not familiar with iptables, but from output of `iptables -t nat -L -v -n`, I can see that traffic from the android to goo.gl is matched by `SS_SPEC_WAN_AC` and `SS_SPEC_WAN_FW`(pks counter increased after request to goo.gl), but the pks counter of the `MASQUARADE` rule in `POSTROUTING` remained the same.

### What is your config in detail (with all sensitive info masked)?

config.json to ss-redir is fairly standard:

    {
      "server": "xxx",
      "server_port": 8443,
      "local_port": 1080,
       "password": "blah",
       "timeout": 300,
       "method": "aes-256-cfb"
    }

**Comment by madeye** [comment]
You should post all the output of `iptables -t nat -L -n`. Otherwise, no one can help.

**Comment by blurrcat** [comment]
@madeye here it is:

    [root@bogon ~]# iptables -L -t nat -v -n
     Chain PREROUTING (policy ACCEPT 8283 packets, 508K bytes)
     pkts bytes target     prot opt in     out     source               destination
        2   128 SS_SPEC_WAN_AC  tcp  --  eth0   *       0.0.0.0/0            0.0.0.0/0
     9918  594K PREROUTING_direct  all  --  *      *       0.0.0.0/0            0.0.0.0/0
     9918  594K PREROUTING_ZONES_SOURCE  all  --  *      *       0.0.0.0/0            0.0.0.0/0
     9918  594K PREROUTING_ZONES  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    
    Chain INPUT (policy ACCEPT 1074 packets, 72346 bytes)
     pkts bytes target     prot opt in     out     source               destination
    
    Chain OUTPUT (policy ACCEPT 1428 packets, 92554 bytes)
     pkts bytes target     prot opt in     out     source               destination
      190 11708 SS_SPEC_WAN_AC  tcp  --  *      *       0.0.0.0/0            0.0.0.0/0
     1605  105K OUTPUT_direct  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    
    Chain POSTROUTING (policy ACCEPT 1521 packets, 98134 bytes)
     pkts bytes target     prot opt in     out     source               destination
     5350  348K MASQUERADE  all  --  *      *       10.42.0.0/24        !10.42.0.0/24
     1701  111K POSTROUTING_direct  all  --  *      *       0.0.0.0/0            0.0.0.0/0
     1701  111K POSTROUTING_ZONES_SOURCE  all  --  *      *       0.0.0.0/0            0.0.0.0/0
     1701  111K POSTROUTING_ZONES  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    
    Chain OUTPUT_direct (1 references)
     pkts bytes target     prot opt in     out     source               destination
    
    Chain POSTROUTING_ZONES (1 references)
     pkts bytes target     prot opt in     out     source               destination
        4   752 POST_public  all  --  *      wlan0   0.0.0.0/0            0.0.0.0/0           [goto]
     1601  105K POST_public  all  --  *      eth0    0.0.0.0/0            0.0.0.0/0           [goto]
       96  5760 POST_public  all  --  *      *       0.0.0.0/0            0.0.0.0/0           [goto]
    
    Chain POSTROUTING_ZONES_SOURCE (1 references)
     pkts bytes target     prot opt in     out     source               destination
    
    Chain POSTROUTING_direct (1 references)
     pkts bytes target     prot opt in     out     source               destination
    
    Chain POST_public (3 references)
     pkts bytes target     prot opt in     out     source               destination
     1701  111K POST_public_log  all  --  *      *       0.0.0.0/0            0.0.0.0/0
     1701  111K POST_public_deny  all  --  *      *       0.0.0.0/0            0.0.0.0/0
     1701  111K POST_public_allow  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    
    Chain POST_public_allow (1 references)
     pkts bytes target     prot opt in     out     source               destination
    
    Chain POST_public_deny (1 references)
     pkts bytes target     prot opt in     out     source               destination
    
    Chain POST_public_log (1 references)
     pkts bytes target     prot opt in     out     source               destination
    
    Chain PREROUTING_ZONES (1 references)
     pkts bytes target     prot opt in     out     source               destination
     8961  547K PRE_public  all  --  wlan0  *       0.0.0.0/0            0.0.0.0/0           [goto]
      957 47006 PRE_public  all  --  eth0   *       0.0.0.0/0            0.0.0.0/0           [goto]
        0     0 PRE_public  all  --  *      *       0.0.0.0/0            0.0.0.0/0           [goto]
    
    Chain PREROUTING_ZONES_SOURCE (1 references)
     pkts bytes target     prot opt in     out     source               destination
    
    Chain PREROUTING_direct (1 references)
     pkts bytes target     prot opt in     out     source               destination
    
    Chain PRE_public (3 references)
     pkts bytes target     prot opt in     out     source               destination
     9918  594K PRE_public_log  all  --  *      *       0.0.0.0/0            0.0.0.0/0
     9918  594K PRE_public_deny  all  --  *      *       0.0.0.0/0            0.0.0.0/0
     9918  594K PRE_public_allow  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    
    Chain PRE_public_allow (1 references)
     pkts bytes target     prot opt in     out     source               destination
    
    Chain PRE_public_deny (1 references)
     pkts bytes target     prot opt in     out     source               destination
    
    Chain PRE_public_log (1 references)
     pkts bytes target     prot opt in     out     source               destination
    
    Chain SS_SPEC_WAN_AC (2 references)
     pkts bytes target     prot opt in     out     source               destination
       97  6136 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0            match-set ss_spec_wan_ac dst
      284 22971 SS_SPEC_WAN_FW  all  --  *      *       0.0.0.0/0            0.0.0.0/0
    
    Chain SS_SPEC_WAN_FW (1 references)
     pkts bytes target     prot opt in     out     source               destination
      284 22971 REDIRECT   tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            redir ports 1080
    
I didn't manully add any rule here. It's basically the default rules + ss-nat rules + hotspot NAT rule.
The ipset `ss_spec_wan_ac` contains local and china ips.

**Comment by madeye** [comment]
You're applying NAT rules on `eth0` interface. But according to your description, you'd like to redirect traffic on `wlan0`.

**Comment by blurrcat** [comment]
I did try ss-nat with `-I wlan0`, it did not work. 
With `-I wlan 0`, here's the change of packets count if I make a request from Android.
Before the request:

    pkts bytes target     prot opt in     out     source               destination
    94  5640 SS_SPEC_WAN_AC  tcp  --  wlan0  *       0.0.0.0/0            0.0.0.0/0
    5448  354K MASQUERADE  all  --  *      *       10.42.0.0/24        !10.42.0.0/24
    74  4440 REDIRECT   tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            redir ports 1080

After the request:

    pkts bytes target     prot opt in     out     source               destination
    96  5760 SS_SPEC_WAN_AC  tcp  --  wlan0  *       0.0.0.0/0            0.0.0.0/0
    5448  354K MASQUERADE  all  --  *      *       10.42.0.0/24        !10.42.0.0/24
    76  4560 REDIRECT   tcp  --  *      *       0.0.0.0/0            0.0.0.0/0            redir ports 1080

As you can see, the packets got matched(94 -> 96) and are redirected to 1080(74 -> 76), but the MASQUERADE seems not to take effect(5448 -> 5448). I suspect that the packets got swallowed by ss-redir and never had a chance to be masqueraded.

**Comment by madeye** [comment]
Double check your config of ss-redir. NAT works as expected according to your info.

**Comment by blurrcat** [comment]
After further debugging, I found that traffic from 10.42.0.0/24 were indeed REDIRECTED, but they are redirected to 10.42.0.1:1080 instead of 127.0.0.1:1080. Here's the log:

    Dec 23 11:31:41 bogon kernel: filter INPUT: IN=wlan0 OUT= MAC=b8:27:eb:7f:36:03:94:65:2d:e1:7f:31:08:00 SRC=10.42.0.26 DST=10.42.0.1 LEN=60 TOS=0x00 PREC=0x00 TTL=64 ID=27368 DF PROTO=TCP SPT=49169 DPT=1080 WINDOW=65535 RES=0x00 SYN URGP=0

Since ss-redir only listens on 127.0.0.1:1080, the packet never reaches ss-redir. It seems there's no option to make ss-redir to listen on other interfaces than lo. Any advice?

**Comment by blurrcat** [comment]
After further debugging, I found that traffic from 10.42.0.0/24 were indeed REDIRECTED, but they are redirected to 10.42.0.1:1080 instead of 127.0.0.1:1080. Here's the log:

    Dec 23 11:31:41 bogon kernel: filter INPUT: IN=wlan0 OUT= MAC=b8:27:eb:7f:36:03:94:65:2d:e1:7f:31:08:00 SRC=10.42.0.26 DST=10.42.0.1 LEN=60 TOS=0x00 PREC=0x00 TTL=64 ID=27368 DF PROTO=TCP SPT=49169 DPT=1080 WINDOW=65535 RES=0x00 SYN URGP=0

Since ss-redir only listens on 127.0.0.1:1080, the packet never reaches ss-redir. It seems there's no option to make ss-redir to listen on other interfaces than lo. Any advice?

**Comment by madeye** [comment]
Look into the document about how to bind to specific address.

**Comment by blurrcat** [comment]
Thanks @madeye, I finally come to the bottom of it.

For those who may be interested, here are the missing links:
* use `ss-nat -I wlan0`, or whatever name of your wifi connection
* iptables would REDIRECT traffic from wifi to 1080. According to [this](https://www.frozentux.net/iptables-tutorial/chunkyhtml/x4529.html), REDIRECT "rewrites the destination address to our own host".
* to my surprise the `dest` is rewriten to `10.42.0.1`(address of wlan0) instead of `127.0.0.1`. The default rules in `filter INPUT` only ACCEPT traffic from interface lo so the redirected traffic from wlan0 is REJECTed.

To summarize, here are the changes I made:
* ss-nat traffic from wlan0: `ss-nat ... -I wlan0`
* accept INPUT traffic from wlan0: `iptables -I INPUT -i wlan0 -j ACCEPT`
* make sure `ss-redir` listens on 10.42.0.1 and localhost. Or simply `ss-redir ... -b 0.0.0.0`
